### DISCUSSION TITLE
How to exclude specific packages from the toolchain?

### DISCUSSION THREAD

**Discussion by netcorefan1** [microsoft/vcpkg#30497] [Q&A] [answered]
Hello,
I've recently integrated in my project vcpkg toolchain through `CMAKE_TOOLCHAIN_FILE` and I really like how all dependencies are automatically linked. It is saving me lot of work!
However, the compilation fails because there are a couple of libraries that needs to be handled from the underline build system instead of the vcpkg toolchain.
I would like to know if there is an easy way to exclude the related packages from being handled by the vcpkg toolchain.
Thanks

**Comment by dg0yt** [comment]
You need to provide empty "overlay" ports for these port names. Empy means that the build recipe does nothing but
~~~cmake
set(VCPKG_POLICY_EMPTY_PACKAGE enabled)
~~~
Your mileage may vary: The remaining vcpkg port may rely on modifications which are absent for system packages.

**Comment by netcorefan1** [comment]
Thanks for your help. Let me understand well...

Supposed that the configuration process ends up with `CMAKE_FREETYPE_DIR=C:/dev/vcpkg/installed/x64-windows/share/freetype` and that `freetype` is the package I want to exclude.
I should add `set(VCPKG_POLICY_EMPTY_PACKAGE enabled)` at the end of the file `C:\dev\vcpkg\ports\freetype\portfile.cmake`, right?
- Is this setting guarantee to be preserved across package updates?
- Any way to set this property through a vcpkg command rather than directly modify the file?

If I am not wrong this apply to system wide, meaning that `freetype` will be excluded in each solution on the machine, not just the specific solution I want it excluded. It would a problem if this is the case.

Not sure what you means for "Your mileage may vary...", but to me sounds like "Yes, you can, but by doing this you may face some issues... ". Could you explain better?

**Reply by Neumann-A** [reply]
> , right?

Wrong you replace the whole contents with just `set(VCPKG_POLICY_EMPTY_PACKAGE enabled)`

> will be excluded in each solution on the machine, not just the specific solution I want it excluded.

It will be excluded by everyone using the same portfile (e.g. using an overlay)

> Could you explain better?

vcpkg adjust targets/paths to work. If that is suddenly missing it might break downstream consumers. You can fix it yourself if you use a system dependency but this requires a bit of extra work.

**Reply by netcorefan1** [reply]
Mhhh...it looks more problematic than I thought.
If I have understood well, the change will not affect just my solution, but the whole vcpkg package management. For example, once `freetype` is set to empty, any subsequent `vcpkg install packageThatNeedsFreetype` will fail (or it will install anyway with unexpected side effects later).
What you mean by system dependency?

There must be some less intrusive way to achieve this goal. What you think about manually changing the CMAKE properties like `CMAKE_FREETYPE_DIR=C:/the_folder_that_would_have_been_set_without_toolchain` ?
Can I expect it will not get overriden by vcpkg when there is already an user defined value?

**Reply by dg0yt** [reply]
> Supposed that the configuration process ends up with `CMAKE_FREETYPE_DIR=C:/dev/vcpkg/installed/x64-windows/share/freetype` and that `freetype` is the package I want to exclude. I should add `set(VCPKG_POLICY_EMPTY_PACKAGE enabled)` at the end of the file `C:\dev\vcpkg\ports\freetype\portfile.cmake`, right?
> 
>     * Is this setting guarantee to be preserved across package updates?
> 
>     * Any way to set this property through a vcpkg command rather than directly modify the file?
> 
> 
> If I am not wrong this apply to system wide, meaning that `freetype` will be excluded in each solution on the machine, not just the specific solution I want it excluded. It would a problem if this is the case.

You are not meant to modify a single file in the vcpkg repo. I'm suggesting to use the `--overlay-ports` option to point to a directory with your variant of the port. Alternatively, you can also work with registries.

> Not sure what you means for "Your mileage may vary...", but to me sounds like "Yes, you can, but by doing this you may face some issues... ". Could you explain better?

As  I wrote, some ports may rely on specific patches or wrappers provided by the ports they depend on. Typical problems solved this way:
- fix broken pkg-config or CMake files,
- deal with different artifact names for dynamic vs. static linkage,
- deal with different artifact names for release vs. debug build type,
- deal with transitive usage requirements, in particular linking requirements for static libs, or macro to control dllexport markup.

And you have to be careful with your system libraries' dependencies. If you want to use system freetype, you probably must also use the system libraries for the direct and transitive dependencies of the system freetype (zlib? brotli? png? ...), to avoid misbehaviour at runtime of the resulting app.

**Reply by dg0yt** [reply]
>  For example, once `freetype` is set to empty, any subsequent `vcpkg install packageThatNeedsFreetype` will fail (or it will install anyway with unexpected side effects later).

No. The installed empty overlay port will satisfy vcpgk's dependency resolution. Whether subsequent installations of dependening ports succeed, is up to you: You need to provide the actual build time requirements (header, libs, pc files, cmake files, etc.) e.g. by installing corresponding system packages.

**Reply by netcorefan1** [reply]
> Mhhh...it looks more problematic than I thought. If I have understood well, the change will not affect just my solution, but the whole vcpkg package management. For example, once `freetype` is set to empty, any subsequent `vcpkg install packageThatNeedsFreetype` will fail (or it will install anyway with unexpected side effects later). What you mean by system dependency?
> 
> There must be some less intrusive way to achieve this goal. What you think about manually changing the CMAKE properties like `CMAKE_FREETYPE_DIR=C:/the_folder_that_would_have_been_set_without_toolchain` ? Can I expect it will not get overriden by vcpkg when there is already an user defined value?

Thanks again for your help. I tried with the above approach which seemed easier and less prone to errors:
`-DFREETYPE_DIR=C:/freetype_custom_build`. 
It worked, vcpkg toolchain recognized my user defined directory without overwriting it and the solution compiled just fine.
However, I also got the following warning which make me think I could get into troubles if I ignore it:
```
-- Configuring done
CMake Warning (dev) at C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:632 (_add_library):
  Policy CMP0003 should be set before this line.  Add code such as

    if(COMMAND cmake_policy)
      cmake_policy(SET CMP0003 NEW)
    endif(COMMAND cmake_policy)

  as early as possible but after the most recent call to
  cmake_minimum_required or cmake_policy(VERSION).  This warning appears
  because target "opencv_world" links to some libraries for which the linker
  must search:

    comctl32, gdi32, ole32, setupapi, ws2_32, wsock32, ws2_32, comctl32, ole32
    setupapi, wsock32, user32, gdi32

  and other libraries with known full path:

    C:/Emugcv47_V2/emgucv/build_x86_64/opencv/3rdparty/lib/Debug/ippiwd.lib
    C:/Emugcv47_V2/emgucv/build_x86_64/opencv/3rdparty/ippicv/ippicv_win/icv/lib/intel64/ippicvmt.lib
    C:/Emugcv47_V2/emgucv/build_x86_64/install/lib/libhdf5.lib
    C:/dev/vcpkg/installed/x64-windows-release/lib/gstreamer-1.0.lib
    C:/dev/vcpkg/installed/x64-windows/lib/glib-2.0.lib

  CMake is adding directories in the second list to the linker search path in
  case they are needed to find libraries from the first list (for backwards
  compatibility with CMake 2.4).  Set policy CMP0003 to OLD or NEW to enable
  or disable this behavior explicitly.  Run "cmake --help-policy CMP0003" for
  more information.
Call Stack (most recent call first):
  opencv/cmake/OpenCVUtils.cmake:1560 (add_library)
  opencv/cmake/OpenCVModule.cmake:970 (ocv_add_library)
  opencv/cmake/OpenCVModule.cmake:882 (_ocv_create_module)
  opencv/modules/world/CMakeLists.txt:60 (ocv_create_module)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Generating done
-- Build files have been written to: C:/Emugcv47_V2/emgucv/build_x86_64
```
It didn't specifically refer to freetype, but it did with other libraries for which I have done the same exclusion by manually specifying location. How should I interpret and fix this?

I also tried the approach you suggested me.  In the case of freetype, I uninstalled it and then:
`vcpkg install freetype --overlay-ports=C:\dev\vcpkg\ports-dummy`.
 The `ports-dummy` directory contains a `freetype` folder with inside a `portfile.cmake`.  The file just contain the line `set(VCPKG_POLICY_EMPTY_PACKAGE enabled)`.
Install didn't showed any error, but also anything about my custom port.
I proceeded with the the configuration and vcpkg toolchain applied the same value which was causing the build failure: `CMAKE_FREETYPE_DIR=C:/dev/vcpkg/installed/x64-windows/share/freetype`.
Shouldn't be populated with something like `CMAKE_FREETYPE_DIR=C:\dev\vcpkg\ports-dummy\freetype` to inform that it is pointing to a dummy package that will redirect to default value provided by the underline build system?
I proceeded anyway with the compilation and it failed with the same error I mentioned at the beginning of the discussion. It failed because freetype pointed to vcpkg install rather then my custom build.
I must for sure miss something obvious.
I also got nearly the same warning, but I'm afraid this happened as soon as I linked vcpkg toolchain rather then as a consequence of my exclusions attempts.

**Reply by dg0yt** [reply]
I think you need help with CMake.

> I also tried the approach you suggested me. In the case of freetype, I uninstalled it and then: `vcpkg install freetype --overlay-ports=C:\dev\vcpkg\ports-dummy`. 
> The `ports-dummy` directory contains a `freetype` folder with inside a `portfile.cmake`. The file just contain the line `set(VCPKG_POLICY_EMPTY_PACKAGE enabled)`. 
> Install didn't showed any error, but also anything about my custom port. I proceeded with the the configuration and vcpkg toolchain applied the same value which was causing the build failure: `CMAKE_FREETYPE_DIR=C:/dev/vcpkg/installed/x64-windows/share/freetype`. 
> Shouldn't be populated with something like `CMAKE_FREETYPE_DIR=C:\dev\vcpkg\ports-dummy\freetype` to inform that it is pointing to a dummy package that will redirect to default value provided by the underline build system?

No. The empty port is only to tell *vcpkg* that the `freetype` dependency is resolved when required by other ports.
The overlay port doesn't interact in any way with what is detected when depending ports or your user project try to find freetype. There shouldn't be any reference to vcpkg's freetype. (However, your user project might have cached information such as `CMAKE_FREETYPE_DIR`, so you would need to restart cleanly.)

**Reply by netcorefan1** [reply]
I restarted everything from scratch just be sure that there is no old cached data here and there. This time `CMAKE_FREETYPE_DIR` does not exists at all which makes me believe that it is working.
Is there a way to automatically generate empty overlay ports rather than having to manually create `freetype folder` + `portfile.cmake` each time?
From my project side the exclusion seems to work fine without any interference from vcpkg toolchain. However I'm now wondering how this change could affect other projects and the vcpkg package management itself:
- Supposing that in another project, this time I need freetype lib from vcpkg. How I can tell vcpkg to not use my empty overlay port?
- When I install a package and that package depends on freetype, can I expect that vcpkg will pickup the needed freetype package rather then my empty overlay? Or I should provide additional settings ?
- Both the approaches to exclude a specific vcpkg package from a project worked fine. The manual one (without overlay port) seemed easier because it just need needs to correct cmake variables without touching anything in the vcpkg folder. However, if you suggested to make extra work and use overlay ports there must be a valid reason. What are the advantage of this approach?

**Reply by dg0yt** [reply]
I have difficulties to link these questions to the full picture.
- There is your "user project", there is the vcpkg installation and there is the outside world.
- Your user project has dependencies. vcpkg is one way to provide these dependencies, in a consistent configuration.
  Dependencies which are not provided by vcpkg must come from the outside world.
  However, there shall be only one active provider of each dependency.
- Now you asked to pick some dependencies from the outside world instead of from vcpkg.
  In your user project, it may be as easy as setting some CMake variables, but it could easily get more involved.
  But you must also ensure that vcpkg picks these dependencies from the outside world, so that the vcpkg variant isn't pulled in again silently. The user project's CMake settings (e.g. `CMAKE_FREETYPE_DIR`) aren't seen by the vcpkg world (except for a few variables controlling manifest mode).
  Here, I suggested to us empty overlay ports. (But I also mentioned registries.) 
  To be clear, this already is somwhat advanced usage of vcpkg.

Now you start to ask about another project which would use another configuration... This isn't really a problem if vcpkg configuration is allowed to vary with project.
 - In classic mode, your user project will take what the vcpkg installation provides. But it is possible to work with different vcpkg installations by choosing the desired installation's cmake toolchain file.
 - In manifest mode, the vcpkg installation will be local to your user project and follow your project's manifest and vcpkg configuration.
 - In both modes, you have to be careful with cached CMake settings (e.g. `CMAKE_FREETYPE_DIR`) in your user project.

**Reply by netcorefan1** [reply]
@dg0yt Many thanks again for being so patient and nice. You really deserve some rewards from Microsoft for being so active and helpful in the community.

> I restarted everything from scratch just be sure that there is no old cached data here and there. This time `CMAKE_FREETYPE_DIR` does not exists at all which makes me believe that it is working...

Unfortunately I'm afraid I was wrong and vcpkg is still taking over. With `CMAKE_TOOLCHAIN_FILE`:
```
-- Setting up FREETYPE
-- FREETYPE:   C:/dev/vcpkg/installed/x64-windows/share/freetype (ver 2.12.1)
-- FREETYPE_DIR: C:/dev/vcpkg/installed/x64-windows/share/freetype
-- FREETYPE_LIBRARIES: C:/dev/vcpkg/installed/x64-windows/bin/freetype.dll
```
Without:
```
-- Setting up FREETYPE
-- FREETYPE:   C:/Emugcv47_V2/emgucv_release/build_x86_64/install/lib/cmake/freetype (ver 2.10.4)
-- FREETYPE_DIR: C:/Emugcv47_V2/emgucv_release/build_x86_64/install/lib/cmake/freetype
-- FREETYPE_LIBRARIES: C:/Emugcv47_V2/emgucv_release/build_x86_64/install/lib/freetype.lib
```
It looks like it doesn't even check my empty `C:\dev\vcpkg\ports-dummy\freetype\portfile.cmake`.
I even tried to make it recognize with `-DCMAKE_TOOLCHAIN_FILE:STRING=$buildSystem/vcpkg.cmake -DVCPKG_OVERLAY_PORTS:STRING=$vcpkg_root/ports-dummy/freetype`, but nothing to do. `$env:VCPKG_OVERLAY_PORTS = "C:\dev\vcpkg\ports-dummy\freetype"` didn't worked too.
This is what I get when I run `vcpkg remove freetype && vcpkg install freetype --overlay-ports=C:\dev\vcpkg\ports-dummy\freetype`:
```
The following packages will be removed:
    freetype:x64-windows
Removing 1/1 freetype:x64-windows
Computing installation plan...
The following packages will be built and installed:
    freetype[brotli,bzip2,core,png,zlib]:x64-windows -> 2.12.1#3
Detecting compiler hash for triplet x64-windows...
Installing 1/1 freetype:x64-windows...
Building freetype[brotli,bzip2,core,png,zlib]:x64-windows...
-- Using cached freetype-2.12.1.tar.xz.
-- Cleaning sources at C:/dev/vcpkg/buildtrees/freetype/src/6482de1748-050bffbef4.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source C:/dev/vcpkg/downloads/freetype-2.12.1.tar.xz
-- Applying patch 0003-Fix-UWP.patch
-- Applying patch brotli-static.patch
-- Applying patch bzip2.patch
-- Applying patch fix-exports.patch
-- Applying patch error-strings.patch
-- Using source at C:/dev/vcpkg/buildtrees/freetype/src/6482de1748-050bffbef4.clean
-- Found external ninja('1.11.0').
-- Configuring x64-windows
-- Building x64-windows-dbg
-- Building x64-windows-rel
-- Fixing pkgconfig file: C:/dev/vcpkg/packages/freetype_x64-windows/lib/pkgconfig/freetype2.pc
-- Using cached msys-mingw-w64-i686-pkg-config-0.29.2-3-any.pkg.tar.zst.
-- Using cached msys-mingw-w64-i686-libwinpthread-git-9.0.0.6373.5be8fcd83-1-any.pkg.tar.zst.
-- Using msys root at C:/dev/vcpkg/downloads/tools/msys2/9a1ec3f33446b195
-- Fixing pkgconfig file: C:/dev/vcpkg/packages/freetype_x64-windows/debug/lib/pkgconfig/freetype2.pc
-- Installing: C:/dev/vcpkg/packages/freetype_x64-windows/share/freetype/usage
-- Performing post-build validation
Elapsed time to handle freetype:x64-windows: 7 s
Total install time: 9 s
freetype is compatible with built-in CMake targets:

    find_package(Freetype REQUIRED)
    target_link_libraries(main PRIVATE Freetype::Freetype) # since CMake 3.10
```
Detection status of my custom port is completely absent while in logs posted from people I can clearly the custom port status:
```
Building package sqlite3[core]:x64-windows...
-- Installing port from location: F:\vcpkg\custom-ports\sqlite3
use custom-ports sqlite
-- Using cached F:/vcpkg/src/downloads/sqlite-amalgamation-3310100.zip
```
Is there something else that I should add other than a `portfile.cmake` file containing just `set(VCPKG_POLICY_EMPTY_PACKAGE enabled)` ?
[This](https://github.com/emgucv/emgucv/blob/master/CMakeLists.txt) is the CMakeList.txt used to build.

>     * In classic mode, your user project will take what the vcpkg installation provides. But it is possible to work with different vcpkg installations by choosing the desired installation's cmake toolchain file.
>     * In manifest mode, the vcpkg installation will be local to your user project and follow your project's manifest and vcpkg configuration.
>     * In both modes, you have to be careful with cached CMake settings (e.g. `CMAKE_FREETYPE_DIR`) in your user project.

Manifest mode is where I had success and without recurring to overlays:
```
{
    "$schema": "https://raw.githubusercontent.com/microsoft/vcpkg/master/scripts/vcpkg.schema.json",
    "name": "myproject",
    "version": "1.0.0",
    "dependencies": [
      "libgeotiff",
      "yasm"
    ]
}
```
By placing the above `vcpkg.json` in the source directory, the toolchain effectively honor the dependencies. If I add "freetype" to the list, the library is installed and linked. If I remove it from the list, when I rebuild, the toolchain does not link it (even if it has been previously installed in the `vcpkg_installed` folder). This approach seemed to me much better and reliable.
However there is a problem: 
there are dependencies that takes huge amount of time. Building each time the same dependency for each project is not feasible.
When I want to get an updated version of a library, I use a script of the type `clone repo > build > move updated lib in my repo > delete all`.
The ideal solution for me would be to keep already built dependencies in the main vcpkg folder (`C:\dev\vcpkg`) and use the above `vcpkg.json` to take the dependencies from there if can be satisfied and create separate installed trees when it is not possible (ex: a different version required from the project).
How I could achieve something like that?

**Reply by dg0yt** [reply, marked as answer]
> This is what I get when I run `vcpkg remove freetype && vcpkg install freetype --overlay-ports=C:\dev\vcpkg\ports-dummy\freetype`:
> 
> ```
> The following packages will be removed:
>     freetype:x64-windows
> Removing 1/1 freetype:x64-windows
> Computing installation plan...
> The following packages will be built and installed:
>     freetype[brotli,bzip2,core,png,zlib]:x64-windows -> 2.12.1#3
> ...

Are you sure about this? I tried to reproduce this locally:

~~~
./vcpkg install --overlay-ports /tmp/overlay/freetype freetype:arm64-linux
Computing installation plan...
The following packages will be built and installed:
    freetype[core]:arm64-linux -> 1 -- /tmp/overlay/freetype
...
~~~

Note how the overlay port version and location is printed in the last line.

**Reply by netcorefan1** [reply]
@dg0yt I'm so sorry for getting back with so much delay. I don't know how I may have missed your comment.

> Is there something else that I should add other than a `portfile.cmake` file containing just `set(VCPKG_POLICY_EMPTY_PACKAGE enabled)` ? 
> 

I wrongly assumed that this was enough. Once I added a `vcpkg.json` with `name`, `version` and `dependencies` it worked.
